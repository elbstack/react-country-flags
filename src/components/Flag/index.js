import React, { PureComponent as Component, createElement } from 'react';
import PropTypes from 'prop-types';
import flags from '../../flags/react';

const fetchFlag = async country => {
  if (!flags[country]) return Promise.resolve(null);
  const { default: data } = await flags[country]();
  return data;
};

// thanks to https://github.com/facebook/react/issues/5465#issuecomment-157888325
const makeCancelable = promise => {
  let hasCanceled_ = false;

  const wrappedPromise = new Promise((resolve, reject) => {
    promise.then(
      val => (hasCanceled_ ? reject({ isCanceled: true }) : resolve(val)),
      error => (hasCanceled_ ? reject({ isCanceled: true }) : reject(error))
    );
  });

  return {
    promise: wrappedPromise,
    cancel() {
      hasCanceled_ = true;
    },
  };
};

export const countries = (() => Object.keys(flags))();

export default class Flags extends Component {
  cancelUpdate = () => {};

  state = {
    flag: null,
  };

  static defaultProps = {
    country: null,
    asSquare: false,
  };

  componentDidMount() {
    this.updateFlag();
  }

  componentDidUpdate(prevProps) {
    const { country } = this.props;

    if (country !== prevProps.country) {
      this.updateFlag();
    }
  }

  componentWillUnmount() {
    this.cancelUpdate();
  }

  updateFlag() {
    const { country } = this.props;
    const { promise, cancel } = makeCancelable(fetchFlag(country));
    promise.then(flag => {
      this.setState({ flag });
    });
    this.cancelUpdate = cancel;
  }

  render() {
    const { asSquare, country, ...props } = this.props;
    const { flag } = this.state;
    if (!flag && !asSquare) return null;
    if (!asSquare) return createElement(flag, props);
    return (
      <svg viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" {...props}>
        {!flag ? null : createElement(flag)}
      </svg>
    );
  }
}

Flags.propTypes = {
  country: PropTypes.string,
  asSquare: PropTypes.bool,
};
