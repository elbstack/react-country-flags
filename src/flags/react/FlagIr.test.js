import React from 'react';
import renderer from 'react-test-renderer';
import Flag from './FlagIr';

it('FlagIr renders without crashing', () => {
  const component = renderer.create(<Flag />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
