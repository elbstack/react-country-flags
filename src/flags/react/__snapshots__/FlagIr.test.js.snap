// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlagIr renders without crashing 1`] = `
<svg
  viewBox="0 0 630 360"
  xmlnsXlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M0 0h630v360H0z"
    fill="#da0000"
  />
  <path
    d="M0 0h630v240H0z"
    fill="#fff"
  />
  <path
    d="M0 0h630v120H0z"
    fill="#239f40"
  />
  <g
    transform="translate(8.4 100.4)"
  >
    <g
      id="s8d56c0e8"
    >
      <g
        fill="none"
        id="s451ef49d"
        stroke="#fff"
        strokeWidth={2}
      >
        <path
          d="M0 1h26M1 10V5h8v4h8V5h-5M4 9h2m20 0h-5V5h8m0-5v9h8V0m-4 0v9"
          id="s4b400911"
          transform="scale(1.4)"
        />
        <path
          d="M0 7h9m1 0h9"
          id="s54a2ec65"
          transform="scale(2.8)"
        />
        <use
          xlinkHref="#s54a2ec65"
          y={120}
        />
        <use
          xlinkHref="#s4b400911"
          y={145.2}
        />
      </g>
      <g
        id="s0d3c66bb"
      >
        <use
          x={56}
          xlinkHref="#s451ef49d"
        />
        <use
          x={112}
          xlinkHref="#s451ef49d"
        />
        <use
          x={168}
          xlinkHref="#s451ef49d"
        />
      </g>
    </g>
    <use
      x={168}
      xlinkHref="#s0d3c66bb"
    />
    <use
      x={392}
      xlinkHref="#s8d56c0e8"
    />
  </g>
  <g
    fill="#da0000"
    transform="matrix(45 0 0 45 315 180)"
  >
    <g
      id="s5e2af08d"
    >
      <path
        d="M-.548.836A.912.912 0 0 0 .329-.722 1 1 0 0 1-.548.836"
      />
      <path
        d="M.618.661A.764.764 0 0 0 .422-.74 1 1 0 0 1 .618.661M0 1l-.05-1L0-.787a.31.31 0 0 0 .118.099V-.1l-.04.993zM-.02-.85L0-.831a.144.144 0 0 0 .252-.137A.136.136 0 0 1 0-.925"
      />
    </g>
    <use
      transform="scale(-1 1)"
      xlinkHref="#s5e2af08d"
    />
  </g>
</svg>
`;
